src/Workspaces/Core/Portable/FindSymbols/SymbolFinder_Helpers.cs

```diff
@@ -196,50 +146,70 @@ private static bool IsAccessible(ISymbol symbol)
+            if (type2Compilation == null)
+                return false;
+
+            // Cache the compilation so that if we need it while checking another set of forwarded types, we don't
+            // expensively throw it away and recreate it.
+            compilationSet.Add(type2Compilation);
+
+            var type1FullMetadataName = type1.ContainingNamespace != null
+                ? $"{type1.ContainingNamespace.ToDisplayString(SymbolDisplayFormats.SignatureFormat)}.{type1.MetadataName}"
+                : type1.MetadataName;
+
+            // Now, find the corresponding reference to type1's assembly in type2's compilation and see if that assembly
+            // contains a forward that matches type2.  If so, type1 was forwarded to type2.
+            var type1AssemblyName = type1.ContainingAssembly.Name;
+            foreach (var assembly in type2Compilation.GetReferencedAssemblySymbols())
             {
-                if (!verifiedKeys.Contains(kvp.Key))
+                if (assembly.Name == type1AssemblyName)
                 {
-                    INamedTypeSymbol originalType, expectedForwardedType;
-                    if (isSearchSymbolCompilation)
-                    {
-                        originalType = kvp.Value.OriginalDefinition;
-                        expectedForwardedType = kvp.Key.OriginalDefinition;
-                    }
-                    else
-                    {
-                        originalType = kvp.Key.OriginalDefinition;
-                        expectedForwardedType = kvp.Value.OriginalDefinition;
-                    }
-
-                    foreach (var referencedAssembly in referencedAssemblies[originalType.ContainingAssembly.Name])
-                    {
-                        var fullyQualifiedTypeName = originalType.MetadataName;
-                        if (originalType.ContainingNamespace != null)
-                        {
-                            fullyQualifiedTypeName = originalType.ContainingNamespace.ToDisplayString(SymbolDisplayFormats.SignatureFormat) +
-                                "." + fullyQualifiedTypeName;
-                        }
-
-                        // Resolve forwarded type and verify that the types from different assembly are indeed equivalent.
-                        var forwardedType = referencedAssembly.ResolveForwardedType(fullyQualifiedTypeName);
-                        if (Equals(forwardedType, expectedForwardedType))
-                        {
-                            verifiedKeys.Add(kvp.Key);
-                            verifiedCount++;
-                        }
-                    }
+                    var forwardedType = assembly.ResolveForwardedType(type1FullMetadataName);
+                    if (Equals(forwardedType, type2))
+                        return true;
                 }
             }
 
-            return verifiedCount;
+            return false;
```

Comment:
core algorithm is *much* simpler, and is effectively jsut asking "is there a TypeForwardedTo entry for this type in the same assembly in the other compilation".  I honestly can't gauge what the old algorithm was doing.  But whatever it was was bogus :)